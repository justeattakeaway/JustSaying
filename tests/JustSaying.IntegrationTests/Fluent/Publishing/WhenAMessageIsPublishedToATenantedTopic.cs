using JustSaying.Extensions;
using JustSaying.Messaging;
using JustSaying.Messaging.MessageHandling;
using JustSaying.Models;
using JustSaying.TestingFramework;
using Microsoft.Extensions.DependencyInjection;
using Newtonsoft.Json;

namespace JustSaying.IntegrationTests.Fluent.Publishing;

public class WhenAMessageIsPublishedToATenantedTopic(ITestOutputHelper outputHelper) : IntegrationTestBase(outputHelper)
{
    [AwsFact]
    public async Task Then_The_Message_Is_Handled()
    {
        // Arrange
        var handler = new InspectableHandler<SimpleMessage>();

        var testId = Guid.NewGuid().ToString("n");

        var topicNameTemplate = "{tenant}-tenanted-topic".TruncateTo(60);

        var services = GivenJustSaying()
            .ConfigureJustSaying((builder) => builder
                .Publications(pub => pub.WithTopic<SimpleMessage>(c =>
                    c.WithTopicName(msg => topicNameTemplate.Replace("{tenant}", msg.Tenant))))
                .Subscriptions(sub =>
                    sub.ForTopic<SimpleMessage>(c => c.WithTopicName("uk-tenanted-topic").WithQueueName($"uk-queue-{testId}"))
                       .ForTopic<SimpleMessage>(c => c.WithTopicName("it-tenanted-topic").WithQueueName($"it-queue-{testId}"))
                       .ForTopic<SimpleMessage>(c => c.WithTopicName("es-tenanted-topic").WithQueueName($"es-queue-{testId}")))
            )
            .AddSingleton<IHandlerAsync<SimpleMessage>>(handler);

        Message CreateMessage(string tenant) => new SimpleMessage()
        {
            Content = testId,
            Tenant = tenant
        };

        string json = string.Empty;

        await WhenAsync(
            services,
            async (publisher, listener, cancellationToken) =>
            {
                await listener.StartAsync(cancellationToken);
                await publisher.StartAsync(cancellationToken);

                // Act
                await publisher.PublishAsync(CreateMessage("uk"), cancellationToken);
                await publisher.PublishAsync(CreateMessage("uk"), cancellationToken);
                await publisher.PublishAsync(CreateMessage("es"), cancellationToken);
                await publisher.PublishAsync(CreateMessage("es"), cancellationToken);
                await publisher.PublishAsync(CreateMessage("it"), cancellationToken);
                await publisher.PublishAsync(CreateMessage("it"), cancellationToken);

                var publisherJson = JsonConvert.SerializeObject(publisher.Interrogate(), Formatting.Indented);

                json = publisherJson.Replace(UniqueName, "integrationTestQueueName", StringComparison.Ordinal);

                // Assert
                await Patiently.AssertThatAsync(OutputHelper,
                    () =>
                    {
                        var received = handler.ReceivedMessages;
                        received.ShouldContain(x => x.Content == testId && x.Tenant == "uk", 2);
                        received.ShouldContain(x => x.Content == testId && x.Tenant == "it", 2);
                        received.ShouldContain(x => x.Content == testId && x.Tenant == "es", 2);
                    });
            });

        json.ShouldMatchApproved(opt => opt.SubFolder("Approvals"));
    }

    [AwsFact]
    public async Task Then_The_Message_Is_Handled_For_Batch()
    {
        // Arrange
        var handler = new InspectableHandler<SimpleMessage>();

        var testId = Guid.NewGuid().ToString("n");

        var topicNameTemplate = "{tenant}-tenanted-topic".TruncateTo(60);

        var services = GivenJustSaying()
            .ConfigureJustSaying((builder) => builder
                .Publications(pub => pub.WithTopic<SimpleMessage>(c =>
                    c.WithTopicName(msg => topicNameTemplate.Replace("{tenant}", msg.Tenant))))
                .Subscriptions(sub =>
                    sub.ForTopic<SimpleMessage>(c => c.WithTopicName("uk-tenanted-topic").WithQueueName($"uk-queue-{testId}"))
                       .ForTopic<SimpleMessage>(c => c.WithTopicName("it-tenanted-topic").WithQueueName($"it-queue-{testId}"))
                       .ForTopic<SimpleMessage>(c => c.WithTopicName("es-tenanted-topic").WithQueueName($"es-queue-{testId}")))
            )
            .AddSingleton<IHandlerAsync<SimpleMessage>>(handler);

        Message CreateMessage(string tenant) => new SimpleMessage()
        {
            Content = testId,
            Tenant = tenant
        };

        string json = string.Empty;

        await WhenBatchAsync(
            services,
            async (publisher, listener, cancellationToken) =>
            {
                await listener.StartAsync(cancellationToken);
                await publisher.StartAsync(cancellationToken);

                // Act
                await publisher.PublishAsync(
                    [
                        CreateMessage("uk"),
                        CreateMessage("uk"),
                        CreateMessage("es"),
                        CreateMessage("es"),
                        CreateMessage("it"),
                        CreateMessage("it"),
                    ],
                    cancellationToken);

                var publisherJson = JsonConvert.SerializeObject(publisher.Interrogate(), Formatting.Indented);

                json = publisherJson.Replace(UniqueName, "integrationTestQueueName", StringComparison.Ordinal);

                // Assert
                await Patiently.AssertThatAsync(OutputHelper,
                    () =>
                    {
                        var received = handler.ReceivedMessages;
                        received.ShouldContain(x => x.Content == testId && x.Tenant == "uk", 2);
                        received.ShouldContain(x => x.Content == testId && x.Tenant == "it", 2);
                        received.ShouldContain(x => x.Content == testId && x.Tenant == "es", 2);
                    });
            });

        json.ShouldMatchApproved(opt => opt.SubFolder("Approvals"));
    }
}
