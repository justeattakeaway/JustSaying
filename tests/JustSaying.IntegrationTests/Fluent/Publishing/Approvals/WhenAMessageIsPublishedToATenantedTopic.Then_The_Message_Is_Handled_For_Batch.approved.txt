{
  "Region": "eu-west-1",
  "Middleware": {
    "Middlewares": []
  },
  "PublishedMessageTypes": {
    "SimpleMessage": {
      "Publishers": {},
      "BatchPublishers": {
        "es-tenanted-topic": {
          "Arn": "arn:aws:sns:us-east-1:000000000000:es-tenanted-topic"
        },
        "it-tenanted-topic": {
          "Arn": "arn:aws:sns:us-east-1:000000000000:it-tenanted-topic"
        },
        "uk-tenanted-topic": {
          "Arn": "arn:aws:sns:us-east-1:000000000000:uk-tenanted-topic"
        }
      }
    }
  },
  "SubscriptionGroups": {
    "Groups": []
  }
}